### ISSUE TITLE
Provide Braille description of emoji

### ISSUE THREAD

**Issue by ABuffEr** [nvaccess/nvda#9213]
Hi,
as you know, NVDA is now able to read CLDR emoji (a great, great feature for social communications).
But, actually, Braille shows only hexadecimal values of these emoji; quite useless if you, temporarily or usually, use Braille without voice (or without hearing it).
So, I would suggest an option to give to the user the possibility to have the spoken emoji description also reported in Braille, perhaps among square brackets or similar.

**Comment by ABuffEr** [comment]
Hi DrSoom 
thanks for links. I searched before opening issue, but I'm not familiar with liblouis system, so I'm not sure about solution discussed, how it'll be shown to normal users, etc. So, regardless of technical difficulties/limitations, I hope those issues can bring to a solution quite similar to one I suggested.

**Comment by DrSooom** [comment]
@ABuffEr: Not really, it only shortens the amount of cells to display the hexadecimal value of undefined Unicode characters. Currently you can save ten cells per character for the 80 Emoticons from [U+1F600 to U+1F64F](https://www.compart.com/en/unicode/block/U+1F600), because of the surrogates in UTF-16. I already uploaded two test files for demonstrating this solution. You can find them there.

Providing a description for Unicode characters isn't part of those two issues, because it's at a completely other level. So your issue here is still valid of course.

But one very important point at the end: The description replacements must be able to enable and disable by the end user, so he is still able to read the hexadecimal values of a Unicode character, if he prefer this type of reading – like I do.

**Comment by DrSooom** [comment]
One more question: Do you want that `😀` would be displayed as `:grinning face:` instead of ⣥⣆⡉⣥⢂⣺ (later on only ⣭⡤⣺) on a braille display?

I guess replacing the characters with their related text replacements in the "cldr.dic" should be not so difficult to implement – always as an option. But on the other hand `:kissing face with smiling eyes:` is quite a lot of text on a braille display. So a hotkey for this new feature to toggle it on and off should be implemented as well at the same time.

**Comment by Andre9642** [comment]
I try to work around that in recent times but currently I haven't found a good solution.
Notwithstanding, the BrailleExtender add-on proposes a small solution.
First, assign a gesture for this command: go to "Input gestures..." dialog, BrailleExtender section then "Show the output speech for selected text in braille".
Then you just need to select text in question and press the defined gesture. The output in braille should display the emoji in plain text.
Note that sometimes it's not necessary to select text.

**Comment by ABuffEr** [comment]
Hi,
first, sorry for delay.
@DrSooom : briefly, yes, I'd like to read :grinning face: (or everything is in CLDR descriptions), and I agree with you about text length, so I agree also it should be an optional feature and controlled by gesture.
@Andre9642 , thanks, I read about your add-on and I'm sure it's a valuable work, I'll try it.

**Comment by DrSooom** [comment]
This issue has been solved by the NVDA add-on [BrailleExtender](https://andreabc.net/projects/NVDA_addons/BrailleExtender/) 20.05.28 and higher. Many thanks to @Andre9642 for implementing this feature in BrailleExtender. Now we will see how long it will take until this and other useful BrailleExtender features are going to be migrated to the NVDA core.

**Comment by leonardder** [comment]
@Andre9642 could you elaborate on how this was fixed in braille extender?

**Comment by ABuffEr** [comment]
Hi all,
I installed the most recent stable version of BrailleExtender. Perfect! It does exactly what I asked for when I opened the issue. Many thanks to @Andre9642  for work and @DrSooom  for reporting.
For me, the issue can be closed, but I could let it open waiting for a core integration.

**Comment by Adriani90** [comment]
Usually issues should not be closed if there is an addon acomplishing what is expected. And specifically in this case, this feature is very important for the core, i.e. for deafblind people for example. So having an addon does not always mean we can close an issue.
I really hope that @Andre9642 shares with us his approach to solve this and maybe donate also code to the NVDA core.
Please always keep in mind that there are many many people out there using NVDA in developing countries where the internet connection is very bad or does not exist at all. Such people would benefit alot from features in the NVDA core and they are not as flexibel to download addons.

**Comment by rperez030** [comment]
Not only that. When you work in a corporate environment behind a firewall sometimes you are not even allowed to install add-ons or they have to undergo an approval process. I agree that having that functionality as part of an add-on is not the solution.

**Comment by DrSooom** [comment]
@leonardder: Start with [undefinedChars.py](https://github.com/Andre9642/BrailleExtender/blob/master/addon/globalPlugins/brailleExtender/undefinedChars.py), [huc.py](https://github.com/Andre9642/BrailleExtender/blob/master/addon/globalPlugins/brailleExtender/huc.py) and [patchs.py](https://github.com/Andre9642/BrailleExtender/blob/master/addon/globalPlugins/brailleExtender/patchs.py). Please note that issue #8702 and this one here were solved together.

**Comment by Andre9642** [comment]
Honestly, save your time! Don't bother reading my code to implement this feature in NVDA core. Better to restart from scratch, IMHO. Indeed:

1. I'm not very very happy of my current implementation. It works but it's probably not efficient/optimized.
2. My code is not commented/documented (I know; it's not a good practice!).
3. I'll make some changes to implement other upcoming features in the add-on (such as the report of attributes via customizable tags and the report of text alignments).

In short, my implementation is not serious for now. Also, I'm not sure to be competent enough to integrate properly a such feature in NVDA Core, at least on the short term.

However, I'm very glad I could help some users with my add-on about this request, it's what counts for me. ☺

**Comment by nvdaes** [comment]
Hello, @ABuffEr and others. This is been discussed now on a spanish mailing list. A person is proposing, at least, showing a unique cell with all dots, so the available script to show the description of the current symbol available in NVDA's core can be used. What do you think? This is triaged with p3 and someone may work for this.

**Comment by ABuffEr** [comment]
Hi @nvdaes,
interesting for quick reading, but sincerely I prefer always the full representation like BrailleExtender does. And I doubt that the work to realize that proposal would be so much different to full representation in terms of Braille spacing management, at code level.

**Comment by nvdaes** [comment]
@ABuffEr , thanks for your comment. I had thought about adding code to the louisHelper translate function, making some replacements if ord(char) is greater than 255. I don't know if this would be the right way.
I saw brailleExtender and seems that it included different options, like the full description, 8 dots etc. I guess that this is controversial and for this reason perhaps it will remain unresolved during a lot of time.

**Comment by ABuffEr** [comment]
> I had thought about adding code to the louisHelper translate function, making some replacements if ord(char) is greater than 255. I don't know if this would be the right way.

Hmm, I never studied NVDA Braille module system. `>255` sounds quite drastic, considering also other languages and Unicode translation feature. Maybe a CLDR based filter would be better.
Just a thought.

**Comment by CoBC** [comment]
Hi,
I also prefer a full representation, it's, IMO, more practice to have the full representation instead of just one full cell and have to press a command all the time to know the character.
Very nice if someone can work on it, BrailleExtender is unfortunately abandoned and this feature of characters description is probably one of the best of this add-on.
Thanks
